File: newcore/libs/jsi/tests/check_test.cpp
#include <Catch2/catch.hpp>
#include <aardvark_jsi/check.hpp>
#include <aardvark_jsi/jsi.hpp>

#ifdef ADV_JSI_JSC
#include <aardvark_jsi/jsc.hpp>
#endif

#ifdef ADV_JSI_QJS
#include <aardvark_jsi/qjs.hpp>
#endif

using namespace aardvark::jsi;

void require_error(
    const std::string& type, const Context& ctx, const Value& value) {
    auto result = check_type(ctx, value, type, {"kind", "name", "target"});
    REQUIRE(result.has_value());
}

void require_valid(
    const std::string& type, const Context& ctx, const Value& value) {
    auto result = check_type(ctx, value, type, {"kind", "name", "target"});
    REQUIRE(!result.has_value());
}

TEMPLATE_TEST_CASE(
    "check",
    "[check]"
#ifdef ADV_JSI_QJS
    ,
    Qjs_Context
#endif
#ifdef ADV_JSI_JSC
    ,
    Jsc_Context
#endif
) {
    auto create_context = []() { return TestType::create(); };
    auto ctx = create_context();

    require_valid("number", *ctx.get(), ctx->value_make_number(21));
    require_error("number", *ctx.get(), ctx->value_make_bool(true));

    auto obj = ctx->object_make(nullptr).to_value();
    require_valid("object", *ctx.get(), obj);
    require_error("array", *ctx.get(), obj);

    auto arr = ctx->object_make_array().to_value();
    require_valid("object", *ctx.get(), arr);
    require_valid("array", *ctx.get(), arr);
    require_error("number", *ctx.get(), arr);
}

File: rust/crates/jsi/tests/check_test.rs
use aardvark_jsi as jsi;
use aardvark_jsi::Context;

fn require_result(
    checker: &dyn jsi::Checker, ctx: &dyn jsi::Context, val: &jsi::Value,
    res: bool,
) {
    let err_params = jsi::CheckErrorParams {
        kind: "kind".to_string(),
        name: "name".to_string(),
        target: "target".to_string(),
    };
    let result = checker.check(ctx, val, &err_params);
    if res {
        assert!(result.is_ok());
    } else {
        assert!(result.is_err());
    }
}

fn require_err(
    checker: &dyn jsi::Checker, ctx: &dyn jsi::Context, val: &jsi::Value,
) {
    require_result(checker, ctx, val, false);
}

fn require_ok(
    checker: &dyn jsi::Checker, ctx: &dyn jsi::Context, val: &jsi::Value,
) {
    require_result(checker, ctx, val, true);
}

#[test]
fn primitive() {
    let ctx = jsi::QjsContext::new();
    require_ok(
        jsi::NUMBER_CHECKER.as_ref(),
        ctx.as_ref(),
        &ctx.value_make_number(21.0),
    );
    require_err(
        jsi::NUMBER_CHECKER.as_ref(),
        ctx.as_ref(),
        &ctx.value_make_bool(true),
    );
}

#[test]
fn objects() {
    let ctx = jsi::QjsContext::new();
    let obj = ctx.object_make(None).to_value();

    require_ok(jsi::OBJECT_CHECKER.as_ref(), ctx.as_ref(), &obj);
    // object is not array
    require_err(jsi::ARRAY_CHECKER.as_ref(), ctx.as_ref(), &obj);

    let arr = ctx.object_make_array().to_value();
    require_ok(jsi::OBJECT_CHECKER.as_ref(), ctx.as_ref(), &arr);
    // but array is an object
    require_ok(jsi::ARRAY_CHECKER.as_ref(), ctx.as_ref(), &arr);
    require_err(jsi::NUMBER_CHECKER.as_ref(), ctx.as_ref(), &arr);
}
